### PR TITLE
Docs: Python script to validate links in wiki docs

### PR DESCRIPTION
**Pull request by tkoehring**
## Pull Request Description

This program parses through all of the markdown files in the [mpich/doc/wiki/](https://github.com/pmodels/mpich/tree/main/doc/wiki) directory and sub-directories. Web links are checked for a `200` HTTP response and relative links are checked to see if the file exists. Relative github links will not work (`../this/file.c#L0-10`).

Additionally the program will check for any files that are not linked to by any other document, indicating they are not used anywhere.

## TODO
- [x] Check if a page is never linked to.

## Sample Output

Broken Link Output: 
```
==== Broken Links ====
File:  wiki/extra/Running_Tests_On_Remote_Machines.md
    Link:  wiki/extra/Category:Stubs

File:  wiki/how_to/Testing_MPICH.md
    Link:  wiki/how_to/source_code/Coding_Standards.md
...
```

Unlinked Files Output
```
==== Unlinked Files ====

Mailing_lists.md
Running_Tests_On_Remote_Machines.md
Index.md
Measuring_Nemesis_Performance.md
...

```
## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
Could you paste the output of running the script?

**Comment by travisk1154** [comment]
> Could you paste the output of running the script?

Sample output is now in the PR description

File: doc/validate_wiki_links.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+            print("\n")
+    print("\n")
+    
+    if _LINK_COUNT:
+        print("==== Unlinked Files ====\n")
+
+        for f in _ALL_FILES:
+            if _ALL_FILES[f] == 0:
+                print(f)
+        print("\n")
```
Comment (hzhou) [inline doc/validate_wiki_links.py:213]: `print()`

**Review by hzhou — CHANGES_REQUESTED** [review]

File: doc/validate_wiki_links.py (hunk 1 above)
Comment (travisk1154) [reply doc/validate_wiki_links.py]: fixed

File: doc/validate_wiki_links.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    for link in links:
+        if link["Broken"]:
+            print("File: ", link["File"])
+            for b in link["Broken"]:
+                print("    Link: ", b)
+            print("\n")
+    print("\n")
+    
+    if _LINK_COUNT:
+        print("==== Unlinked Files ====\n")
```
[the remark this answers is not in the document]
Comment (travisk1154) [reply doc/validate_wiki_links.py:208]: fixed

File: doc/validate_wiki_links.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        links.append(validate_links(file))
+
+    print("==== Broken Links ====\n")
+
+    for link in links:
+        if link["Broken"]:
+            print("File: ", link["File"])
+            for b in link["Broken"]:
+                print("    Link: ", b)
+            print("\n")
```
[the remark this answers is not in the document]
Comment (travisk1154) [reply doc/validate_wiki_links.py]: fixed

**Comment by hzhou** [comment]
Result:
```
$ python validate_wiki_links.py --unlinked-files
==== Broken Links ====
File:  wiki/extra/Running_Tests_On_Remote_Machines.md
    Link:  wiki/extra/Category:Stubs

File:  wiki/notes/request.md
    Link:  wiki/notes/../../../src/include/mpir_request.h#L168-L225
    Link:  wiki/notes/../../../src/include/mpir_request.h#L168-L225
    Link:  wiki/notes/../../../src/mpid/ch4/include/mpidpre.h#L272-L304
    Link:  wiki/notes/../../../src/mpid/ch4/include/mpidpre.h#L207-L222
    Link:  wiki/notes/../../../src/mpid/ch4/netmod/ofi/ofi_pre.h#L170-L188

File:  wiki/source_code/Contributing_to_MPICH.md
    Link:  wiki/source_code/Version_Control_System.md

File:  wiki/source_code/Managing_Commits_In_Trunk_and_Release_Branches.md
    Link:  wiki/source_code/Git.md

File:  wiki/source_code/Building_MPICH_For_Development.md
    Link:  wiki/source_code/Getting_And_Building_MPICH#Setting_up_the_build_environment
    Link:  wiki/source_code/PE_RTE

File:  wiki/source_code/Git_Usage.md
    Link:  wiki/source_code/SVN

File:  wiki/source_code/Creating_A_Release.md
    Link:  wiki/source_code/../testing/Testing_MPICH.md#running-the-special-tests
    Link:  wiki/source_code/Git.md

File:  wiki/source_code/Github.md
    Link:  wiki/source_code/Version_Control_System.md

File:  wiki/design/Reporting_And_Returning_Error_Codes.md
    Link:  wiki/design/../source_code/Coding_Standards.md#returning-error-codes

File:  wiki/design/Environment_Variables.md
    Link:  wiki/design/Category:Stubs
    Link:  wiki/design/Category:Deprecated

File:  wiki/design/Tool_Interfaces.md
    Link:  wiki/design/Parameters_in_MPICH
    Link:  wiki/design/Internal_Instrumentation
    Link:  wiki/design/Category:Design_Documents

File:  wiki/design/Nemesis_Network_Module_API.md
    Link:  wiki/design/Media:put.fig.txt
    Link:  wiki/design/Media:get.fig.txt
    Link:  wiki/design/Media:coop.fig.txt

File:  wiki/design/RMA_Design.md
    Link:  wiki/design/MPI-3_RMA_Implementation_Timeline

File:  wiki/design/The_Progress_Engine.md
    Link:  wiki/design/Making_MPICH_Thread_Safe.md#the-progress-engine

File:  wiki/how_to/Testing_MPICH.md
    Link:  wiki/how_to/source_code/Coding_Standards.md

File:  wiki/tools/TAU_by_example.md
    Link:  wiki/tools/../text/Wave2d.cpp
    Link:  wiki/tools/HPCToolkit_by_example.md#hardware-counters
    Link:  wiki/tools/HPCToolkit_by_example.md#hardware-counters

File:  wiki/tools/MPE_by_example.md
    Link:  wiki/tools/../text/Ring.c
    Link:  wiki/tools/../text/Makefile.ring
    Link:  wiki/tools/../text/Kmeans.cpp
    Link:  wiki/tools/../text/Makefile.kmeans

File:  wiki/tools/HPCToolkit_by_example.md
    Link:  wiki/tools/../text/Mmult.c

File:  wiki/developer_guide.md
    Link:  wiki/../../src/binding/api_mapping.txt
    Link:  wiki/../../src/include/mpir_objects.h#L138-L155
    Link:  wiki/../../src/mpi/comm/commutil.c#L21
    Link:  wiki/../../src/mpi/comm/commutil.c#L22
    Link:  wiki/../../src/mpid/ch3/include/mpidpre.h#L532-L753
    Link:  wiki/../../src/mpid/ch4/include/mpidch4.h#L15-L313

File:  wiki/media/SC11_BoF.md
    Link:  wiki/media/media:SC11-ANL-Intro.pdf
    Link:  wiki/media/media:SC11-Cray.pdf
    Link:  wiki/media/media:SC11-Microsoft.pdf
    Link:  wiki/media/media:SC11-OSU.pdf
    Link:  wiki/media/media:SC11-IBM.pdf

File:  wiki/testing/Testing_MPICH.md
    Link:  wiki/testing/Coding_Standards

File:  wiki/testing/Nightly_Tests.md
    Link:  wiki/testing/Testing_MPICH.md#running-the-special-tests
    Link:  wiki/testing/Testing_MPICH.md#running-the-configure-tests

==== Unlinked Files ====
Running_Tests_On_Remote_Machines.md
PE_RTE.md
Getting_And_Building_MPICH.md
Git_Usage.md
Index.md
MPICH_Namespacing.md
Environment_Variables.md
MPI-3_RMA_Implementation_Timeline.md
Testing_MPICH.md
Measuring_Nemesis_Performance.md
Student_Projects.md
SC11_BoF.md
Mailing_lists.md
```

**Review by hzhou — APPROVED** [review]
Looks good. Could you rebase the PR? You may merge them after rebasing.

**Comment by hzhou** [comment]
Oh, make `--unlinked-files` default and get rid of the option

**Comment by travisk1154** [comment]
> Oh, make `--unlinked-files` default and get rid of the option

Fixed this and rebased. Good to go.

**Comment by hzhou** [comment]
Looks good, could you rebase one more time?

**Pull request merged**